docs/proposals/device-crd.md

```diff
@@ -89,74 +89,117 @@ A `device model` describes the device properties exposed by the device and prope
 // DeviceModelSpec defines the model / template for a device.It is a blueprint which describes the device
 // capabilities and access mechanism via property visitors.
 type DeviceModelSpec struct {
-  // List of device properties.
-  // Required.
-  DeviceProperties       []DeviceProperty        `json:"properties,omitempty"`
-  // List of property visitors which describe how to access the device properties.
-  // Required.
-  DevicePropertyVisitors []DevicePropertyVisitor `json:"propertyVisitors,omitempty"`
+	// Required: List of device properties.
+	Properties []DeviceProperty `json:"properties,omitempty"`
+	// Required: List of property visitors which describe how to access the device properties.
+	PropertyVisitors []DevicePropertyVisitor `json:"propertyVisitors,omitempty"`
 }
 
 // DeviceProperty describes an individual device property / attribute like temperature / humidity etc.
 type DeviceProperty struct {
-  // The device property name.
-  // Required.
-  Name         string            `json:"name,omitempty"`
-  // The device property description.
-  // +optional
-  Description  string            `json:"description,omitempty"`
-  // Additional configuration about the device property like datatype, range, min, max etc.
-  // Optional: If no configuration details are provided, it is assumed that the property is an empty read-only string.
-  // +optional
-  Config       map[string]string `json:"config,omitempty"`
+	// Required: The device property name.
+	Name string `json:"name,omitempty"`
+	// The device property description.
+	// +optional
+	Description string `json:"description,omitempty"`
+	// Required: PropertyType represents the type and data validation of the property.
+	PropertyType `json:",inline"`
 }
 
-// DevicePropertyVisitor describes the specifics of accessing a particular device
-// property. Visitors are intended to be consumed by device mappers which connect to devices
-// and collect data / perform actions on the device. A device may support multiple protocols.
-// We can define multiple visitors per protocol for a device. For e.g If a device supports Bluetooth
-// protocol , the visitors would describe the 128-bit UUIDs to read / write charactersitics of a Bluetooth
-// service exposed by the device.
-type DevicePropertyVisitor struct {
-  // The protocol type to connect to the device and access the property.
-  // Required.
-  Protocol     ProtocolType      `json:"protocol,omitempty"`
-  // A list of visitors describing the properties which can be accessed with the protocol.
-  // Required.
-  Visitors     []PropertyVisitor `json:"visitors,omitempty"`
+// Represents the type and data validation of a property.
+// Only one of its members may be specified.
+type PropertyType struct {
+	// +optional
+	Int    PropertyTypeInt64
```

Comment:
do we need json descriptors for all the properties ?

Reply:
It's actually not required in a design doc, we can add them in the real code.

Reply:
ok got it